[#6936] improvement(doc): update doc for fileset multiple locations (#7009)
apache/gravitino: docs/manage-fileset-metadata-using-gravitino.md

   specified as `EXTERNAL`, the files of the fileset will **not** be deleted when
   the fileset is dropped.

:::note
If the locations of the manged fileset do not exist, Gravitino will create/delete the locations when the fileset is created/deleted.
Unless the catalog property `disable-filesystem-ops` is set to true or the location contains a [placeholder](./manage-fileset-metadata-using-gravitino.md#placeholder).
:::

#### storageLocation

The `storageLocation` is the physical location of the fileset. Users can specify this location
when creating a fileset, or follow the rules of the catalog/schema location if not specified.

The value of `storageLocation` depends on the configuration settings of the catalog:

> - If this is a local fileset catalog, the `storageLocation` should be in the format of `file:///path/to/fileset`.

Reviewer: ```suggestion
- For a local fileset catalog, the `storageLocation` should be in the format of `file:///path/to/fileset`.
```

> - If this is a HDFS fileset catalog, the `storageLocation` should be in the format of `hdfs://namenode:port/path/to/fileset`.

Reviewer: ```suggestion
- For a HDFS fileset catalog, the `storageLocation` should be in the format of `hdfs://namenode:port/path/to/fileset`.
```

   corresponding fileset property value.
2. When the catalog property `location` is specified but the schema property `location` isn't specified, the storage location is:
   1. `catalog location/schema name/fileset name` if `catalog location` does not contain any placeholder. 
   2. `catalog location` - placeholders in the catalog location will be replaced by the corresponding fileset property value.

3. When the catalog property `location` isn't specified but the schema property `location` is specified,
   the storage location is:
   1. `schema location/fileset name` if `schema location` does not contain any placeholder.
   2. `schema location` - placeholders in the schema location will be replaced by the corresponding fileset property value.
   
4. When both the catalog property `location` and the schema property `location` are specified, the storage
   location is:
   1. `schema location/fileset name` if `schema location` does not contain any placeholder.
   2. `schema location` - placeholders in the schema location will be replaced by the corresponding fileset property value.

5. When both the catalog property `location` and schema property `location` isn't specified, the user

>    should specify the `storageLocation` in the fileset creation.

Reviewer: The logic is not very complicated, but the description is difficult to parse.
Based on the description above, the logic is

- if the `storageLocation` parameter is provided in the fileset creation request,
  we'll use it with placeholders (if any) properly substituted, or
- if `location` is specified in the schema, we use `<schema location>/<fileset name>`,
  where the placeholders (if any) in the `<schema location>` are substituted properly
  using fileset property values. Or,
- if `location` is specified in the catalog, we use `<catalog location>/<schema name>/<fileset name>`,
  where the placeholders (if any) in the `<catalog location>` are substituted properly
  using fileset property values. Or,
- we throw an error for missing storage location information.

Author: Thank you for pointing this out. How about change to:
```markdown
For a `MANAGED` fileset, the storage location is determined in the following priority order:

1. If the user specifies `storageLocation` during fileset creation:
   - This location is used, with any [placeholders](#placeholder) replaced by the corresponding fileset property values.

2. If the user doesn't specify `storageLocation`:
   - If schema property `location` is specified:
     - Use `<schema location>/<fileset name>` if schema location has no placeholders
     - Use `<schema location>` with placeholders replaced by fileset property values
   
   - Otherwise, if catalog property `location` is specified:
     - Use `<catalog location>/<schema name>/<fileset name>` if catalog location has no placeholders
     - Use `<catalog location>` with placeholders replaced by fileset property values
   
   - If neither schema nor catalog location is specified:
     - The user must provide `storageLocation` during fileset creation

For an `EXTERNAL` fileset, the user must always specify `storageLocation` during fileset creation. If the provided location contains placeholders, they will be replaced by the corresponding fileset property values.
```

Reviewer: Yup. The revised version works too.
The key point is to have users easily deduce the behavior once they know the logic behind.

   }
)

# create a fileset by placeholders
catalog.as_fileset_catalog().create_multiple_location_fileset(
    ident=NameIdentifier.of("test_schema", "example_fileset"),
    type=Fileset.Type.MANAGED,
    comment="This is an example fileset",
    storage_locations={
        "l4": "gs://myBucket/{{catalog}}/{{schema}}/workspace_{{project}}/{{user}}",
    },
    roperties={
       "placeholder-project": "test_project",
       "placeholder-user": "test_user",
       "default-location-name": "l1",
    }
)

# the fileset will be created with 4 storage locations:
{
  "name": "example_fileset",
  "comment": "This is an example fileset",
  "type": "MANAGED",
  "storageLocation": null,
  "storageLocations": {
    "l1": "file:///test_catalog/test_schema/workspace_test_project/test_user",
    "l2": "hdfs:///test_catalog/test_schema/workspace_test_project/test_user",
    "l3": "s3a://myBucket/test_catalog/test_schema/workspace_test_project/test_user",
    "l4": "gs://myBucket/test_catalog/test_schema/workspace_test_project/test_user"
  },
  "properties": {
    "placeholder-project": "test_project",
    "placeholder-user": "test_user",
    "default-location-name": "l1"
  }
}
```

</TabItem>

> </Tabs>

Author: GVFS can utilize the configuration described there; should I also add a code example there? 

https://github.com/apache/gravitino/blob/c37e02c21e700b669a29a1d9bd6d13b59f0c67e9/docs/how-to-use-gvfs.md?plain=1#L68-L69

> </Tabs>

Reviewer 2: Yes, the code example of how to use multiple-location fileset by GVFS.

> </Tabs>

Author: added

  }
}' http://localhost:8090/api/metalakes/metalake/catalogs

# create a schema under the catalog
curl -X POST -H "Accept: application/vnd.gravitino.v1+json" \
-H "Content-Type: application/json" -d '{
  "name": "test_schema",
  "comment": "comment",
  "properties": {
    "location-l3": "s3a://myBucket/{{catalog}}/{{schema}}/workspace_{{project}}/{{user}}"
  }
}' http://localhost:8090/api/metalakes/metalake/catalogs/test_catalog/schemas

# create a fileset by placeholders
curl -X POST -H "Accept: application/vnd.gravitino.v1+json" \
-H "Content-Type: application/json" -d '{
  "name": "example_fileset",
  "comment": "This is an example fileset",
  "type": "MANAGED",
  "storageLocations": {
    "l4": "gs://myBucket/{{catalog}}/{{schema}}/workspace_{{project}}/{{user}}"
  },
  "properties": {
    "placeholder-project": "test_project",
    "placeholder-user": "test_user",
    "default-location-name": "l1"
  }
}' http://localhost:8090/api/metalakes/metalake/catalogs/test_catalog/schemas/test_schema/filesets

# the fileset will be created with 4 storage locations:
{
  "name": "example_fileset",
  "comment": "This is an example fileset",
  "type": "MANAGED",
  "storageLocation": null,

>   "storageLocations": {

Reviewer: `storageLocation` is for backward compatibility?

Why do we have both singular and plural forms of the same property?

Can we unify this into just a `storageLocations` map?

Reviewer: Another question ...
Is `storageLocation` a required property?
Why do we need to explicitly set a property that might be defaulted to null?

Author: > `storageLocation` is for backward compatibility?
> 
> Why do we have both singular and plural forms of the same property?
> 
> Can we unify this into just a `storageLocations` map?

yeah, `storageLocation` is for backward compatibility.

Author: > Is `storageLocation` a required property?

No.

> Why do we need to explicitly set a property that might be defaulted to null?

This is not a creation request, but an example of a returned result.

Reviewer: Makes sense.
